controllers/cinderapi_controller.go

```diff
@@ -169,95 +184,349 @@ func (r *CinderAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+	publicEndpointData := common.EndpointData{
+		Port: cinder.CinderPublicPort,
+		Path: "/v2/%(project_id)s",
+	}
+	internalEndpointData := common.EndpointData{
+		Port: cinder.CinderInternalPort,
+		Path: "/v2/%(project_id)s",
+	}
+	data := map[common.Endpoint]common.EndpointData{
+		common.EndpointAdmin:    adminEndpointData,
+		common.EndpointPublic:   publicEndpointData,
+		common.EndpointInternal: internalEndpointData,
+	}
+
+	apiEndpointsV2, ctrlResult, err := common.ExposeEndpoints(
+		ctx,
+		helper,
+		cinder.ServiceName,
+		serviceLabels,
+		data,
+	)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+
+	//
+	// Update instance status with service endpoint url from route host information for v2
+	//
+	// TODO: need to support https default here
+	if instance.Status.APIEndpoints == nil {
+		instance.Status.APIEndpoints = map[string]map[string]string{}
+	}
+	instance.Status.APIEndpoints[cinder.ServiceNameV2] = apiEndpointsV2
+	// V2 - end
+
+	// V3
+	adminEndpointData = common.EndpointData{
+		Port: cinder.CinderAdminPort,
+		Path: "/v3/%(project_id)s",
+	}
+	publicEndpointData = common.EndpointData{
+		Port: cinder.CinderPublicPort,
+		Path: "/v3/%(project_id)s",
+	}
+	internalEndpointData = common.EndpointData{
+		Port: cinder.CinderInternalPort,
+		Path: "/v3/%(project_id)s",
+	}
+	data = map[common.Endpoint]common.EndpointData{
+		common.EndpointAdmin:    adminEndpointData,
+		common.EndpointPublic:   publicEndpointData,
+		common.EndpointInternal: internalEndpointData,
+	}
+
+	apiEndpointsV3, ctrlResult, err := common.ExposeEndpoints(
+		ctx,
+		helper,
+		cinder.ServiceName,
+		serviceLabels,
+		data,
+	)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+
+	//
+	// Update instance status with service endpoint url from route host information for v2
+	//
+	// TODO: need to support https default here
+	instance.Status.APIEndpoints[cinder.ServiceNameV3] = apiEndpointsV3
+	// V3 - end
 
-	// TODO:
-	// get readinessProbes
-	//readinessProbe := util.Probe{ProbeType: "readiness"}
-	//livenessProbe := util.Probe{ProbeType: "liveness"}
+	// expose service - end
 
-	// get volumes
-	initVolumeMounts := common.GetInitVolumeMounts()
-	volumeMounts := common.GetVolumeMounts()
-	volumes := common.GetVolumes(instance.Spec.ManagingCrName)
+	//
+	// create users and endpoints - https://docs.openstack.org/Cinder/latest/install/install-rdo.html#configure-user-and-endpoints
+	// TODO: rework this
+	//
+	ospSecret, _, err := common.GetSecret(ctx, helper, instance.Spec.Secret, instance.Namespace)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			return ctrl.Result{RequeueAfter: time.Second * 10}, fmt.Errorf("OpenStack secret %s not found", instance.Spec.Secret)
+		}
+		return ctrl.Result{}, err
+	}
 
-	deployment := &appsv1.Deployment{
+	cinderKeystoneServiceV2 := &keystonev1beta1.KeystoneService{
```

Comment:
should use new NewKeystoneService lib we now have in lib-common like placement now use it https://github.com/openstack-k8s-operators/placement-operator/pull/17

Reply:
Done